src/hotspot/share/prims/jvm.cpp

```diff
@@ -3282,6 +3282,8 @@ JVM_END
 
 JVM_ENTRY(jboolean, JVM_ReferenceRefersTo(JNIEnv* env, jobject ref, jobject o))
   oop ref_oop = JNIHandles::resolve_non_null(ref);
+  // PhantomReference has it's own implementation of refersTo().
+  assert(!java_lang_ref_Reference::is_phantom(ref_oop), "precondition");
```

Comment:
Maybe instead assertions should be added to the java_lang_ref_Reference functions for accessing the referent?  (weak_referent_no_keepalive, weak_referent, phantom_referent_no_keepalive)

Reply:
I fell like that is some what orthogonal to what this PR is trying to address. This wants to make the precondition clear for the JVM entry API. As a PhantomReference is a Reference it was not clear from JVM_ReferenceRefersTo entry that it is not the entry for PhantomReference. This assert and comment aims to make this clear.

But I do agree that all the accessors in `java_lang_ref_Reference` should assert the type. I will make another issue with the following change.
```diff
oop java_lang_ref_Reference::weak_referent_no_keepalive(oop ref) {
+ assert(!java_lang_ref_Reference::is_phantom(ref_oop), "must be Weak or Soft Reference");
  return ref->obj_field_access<ON_WEAK_OOP_REF | AS_NO_KEEPALIVE>(_referent_offset);
}

oop java_lang_ref_Reference::weak_referent(oop ref) {
+ assert(!java_lang_ref_Reference::is_phantom(ref_oop), "must be Weak or Soft Reference");
  return ref->obj_field_access<ON_WEAK_OOP_REF>(_referent_offset);
}

oop java_lang_ref_Reference::phantom_referent_no_keepalive(oop ref) {
+ assert(java_lang_ref_Reference::is_phantom(ref_oop), "must be Phantom Reference");
  return ref->obj_field_access<ON_PHANTOM_OOP_REF | AS_NO_KEEPALIVE>(_referent_offset);
}
```